agent/dns.go

```diff
@@ -1169,6 +1171,26 @@ func (d *DNSServer) serviceNodeRecords(dc string, nodes structs.CheckServiceNode
 		// Add the node record
 		records := d.formatNodeRecord(node.Node, addr, qName, qType, ttl, edns, true)
 		if records != nil {
+			// only allow at most 1 CNAME record
+			switch records[0].(type) {
+			case *dns.CNAME:
+				if haveCNAME || !allowCNAME {
```

Comment:
My recommendation would be to add some detailed comments here. It took me awhile reading the code to figure out what was going on and I'm questioning whether I even understood it correctly. It'd help to have a 2nd source helping me here.

Reply:
Instead of docs I just refactored it a lot and it should be much more straight forward.

Now we hold onto the first CNAME RR (+ associated RRs) until the end of the loop. If we have nothing in the Answer section then we add the CNAME answer.